cupyx/scipy/linalg/_matfuncs.py

```diff
@@ -35,3 +37,95 @@ def khatri_rao(a, b):
 
     c = a[..., :, cupy.newaxis, :] * b[..., cupy.newaxis, :, :]
     return c.reshape((-1,) + c.shape[2:])
+
+
+# ### expm ###
+b = [64764752532480000.,
+     32382376266240000.,
+     7771770303897600.,
+     1187353796428800.,
+     129060195264000.,
+     10559470521600.,
+     670442572800.,
+     33522128640.,
+     1323241920.,
+     40840800.,
+     960960.,
+     16380.,
+     182.,
+     1.,]
+
+th13 = 5.37
+
+
+def expm(a):
+    """Compute the matrix exponential.
+
+    Parameters
+    ----------
+    a : ndarray, 2D, float64
+
+    Returns
+    -------
+    matrix exponential of `a`
+
+    Notes
+    -----
+    Uses (a simplified) version of Algorithm 2.3 of [1]_:
+    a [13 / 13] Pade approximant with scaling and squaring.
+
+    Simplifications:
+    - we always use a [13/13] approximant
+    - no matrix balancing
+
+    References
+    ----------
+    .. [1] N. Higham, SIAM J. MATRIX ANAL. APPL. Vol. 26(4), p. 1179 (2005)
+       https://doi.org/10.1137/04061101X
+
+    """
+    if a.size == 0:
+        return cupy.zeros((0, 0), dtype=a.dtype)
+
+    n = a.shape[0]
+
+    # try reducing the norm
+    mu = cupy.diag(a).sum() / n
+    A = a - cupy.eye(n)*mu
+
+    # scale factor
+    nrmA = cupy.linalg.norm(A, ord=1)
+
+    scale = nrmA > th13
```

Comment:
Transferring the data of `nrmA` to CPU first would reduce the number of sync. I suppose, ideally, we could eliminate the sync to do these scaling and squaring later with keeping `s` on GPU, but it would include troublesome cuBLAS device API calls and HIP compatibility issues. It would be too much here.

Reply:
Also, can I ask you to have `expm` in another PR to separate from this one?

Reply:
Would it help to keep `b` and scalars s and th13 on the device?

Reply:
I suppose we need `s` on the host to the squaring:
```py
for _ in range(s):
    x = x @ x
```
We could do this repetitive matrix multiplication if we have `s` on the device, but it would require cuBLAS device API calls in a RawKernel.

Reply:
I see. Just confirm my understanding: changing the matmul into `cupy.matmul(x, x, out=x) ` also incurs synchronizarion, correct?

Reply:
Both of the matmul and `cupy.matmul` do not have synchronization. In the current code, `if scale` where `scale` is evaluated as if `bool(scale)` and `float(nrmA)` have.